### PR TITLE
[6.3] Added stubs for TemplateSync - Git Templates Feature

### PR DESCRIPTION
**Pull request by rochacbruno**
Added stubs for UI|API TemplateSync (a.k.a Git Templates) feature
more information on 6.3 feature spreadsheet and /apidoc/v2/template.html
(automation will depend on https://github.com/SatelliteQE/nailgun/issues/435)

### CODE REVIEW

File: tests/foreman/api/test_template.py [hunk 1]
```diff
@@ -305,3 +305,349 @@ def test_positive_clone(self):
...
+class TemplateSyncTestCase(APITestCase):
+    """Implements TemplateSync tests from API"""
+
+    @classmethod
+    def setUpClass(cls):
+        """Setup for TemplateSync functionality
+
+        :steps:
+
+            1. Git repository must exist (in gitlab or github) and its url
```
Comment (oshtaier) [inline tests/foreman/api/test_template.py]: some of these are not actual steps, but more preconditions, so I will put such into `:setup:` section

File: tests/foreman/api/test_template.py [hunk 2]
```diff
@@ -305,3 +305,349 @@ def test_positive_clone(self):
...
+        """Assure only templates with a given filter regex are pulled from
+        git repo.
+
+        :id: 628a95d6-7a4e-4e56-ad7b-d9fecd34f765
+
+        :Steps:
+            1. Using nailgun (pending implementation) or direct API call
+               import only the templates matching with regex e.g: `^atomic.*`
+               refer to: `/apidoc/v2/template/import.html`
+            2. Assert result is {'message': 'success'} and template imported
```
Comment (oshtaier) [inline tests/foreman/api/test_template.py]: Assertion are not actual steps, so I put such into `expectedresults` section

File: tests/foreman/api/test_template.py [hunk 3]
```diff
@@ -305,3 +305,349 @@ def test_positive_clone(self):
...
+            3. Assert locked template is forced to update.
+
+        :expectedresults: Locked template is updated by force.
+
+        :CaseImportance: Critical
+        """
+
+    @stubbed()
+    @tier1
+    def test_positive_import_associated_with_os(self):
```
Comment (oshtaier) [inline tests/foreman/api/test_template.py]: `test_positive_import_associated_with_taxonomies`?

File: tests/foreman/ui/test_template.py [hunk 4]
```diff
@@ -512,3 +512,63 @@ def test_positive_export_template_disable_safe_render(self):
...
+               via ssh://.
+            3. Local directory /var/tmp/templatesync-{random}/ must be created.
+            4. Organization and Location must be created to isolate the
+               templates ownership.
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_positive_templatesync_settings_is_enabled(self):
```
Comment (oshtaier) [inline tests/foreman/ui/test_template.py]: `test_positive_settings_is_enabled` as you already has `TemplateSyncTestCase` class name

File: tests/foreman/ui/test_template.py [hunk 5]
```diff
@@ -512,3 +512,63 @@ def test_positive_export_template_disable_safe_render(self):
...
+
+        :expectedresults: Plugin is installed so TemplateSync tab is present.
+
+        :CaseImportance: Critical
+        """
+
+    @run_only_on('sat')
+    @stubbed()
+    @tier1
+    def test_positive_templatesync_git_settings(self):
```
Comment (oshtaier) [inline tests/foreman/ui/test_template.py]: same here

File: tests/foreman/api/test_template.py [hunk 6]
```diff
@@ -305,3 +305,349 @@ def test_positive_clone(self):
...
+
+    @stubbed()
+    @tier1
+    def test_positive_export_and_import_with_metadata(self):
+        """Assure exported template contains metadata.
+
+        :id: ba8a34ce-c2c6-4889-8729-59714c0a4b19
+
+        :Steps:
+            1. Create a template using nailgun entity and specify Oorg/Loc.
```
Comment (oshtaier) [inline tests/foreman/api/test_template.py]: `Create a template using nailgun entity and specify Org/Loc.`

**Comment by rochacbruno** [comment]
@oshtaier comments addressed

File: tests/foreman/api/test_template.py [hunk 7]
```diff
@@ -305,3 +305,334 @@ def test_positive_clone(self):
                 for key in uniqe_keys:
```
Comment (cswiii) [inline tests/foreman/api/test_template.py]: 'unique_keys', perhaps?  Might not matter much.

Comment (rochacbruno) [inline tests/foreman/api/test_template.py]: @cswiii yes I agree but the code above do not belong to this PR, we can fix it in a different PR maybe.

**Comment by rochacbruno** [comment]
@kbidarkar `foreman-installer --enable-foreman-plugin-templates` must be part of provisioning for this feature to be tested. How can we enable it?

**Comment by rochacbruno** [comment]
Rebased and passed waiting for one more reviewer @SatelliteQE/quality-engineers

File: tests/foreman/api/test_template.py [hunk 8]
```diff
@@ -306,3 +312,334 @@ def test_positive_clone(self):
...
+            1. Using nailgun or direct API call
+               import the templates NOT matching with regex e.g: `^freebsd.*`
+               refer to: `/apidoc/v2/template/import.html` using the
+               {'negate': true} in POST body to negate the filter regex.
+
+        :expectedresults:
+            1. Assert result is {'message': 'success'}
+            2. Assert templates mathing the regex were not pulled.
+
+        :CaseImportance: Critical
```
Comment (sghai) [inline tests/foreman/api/test_template.py]: we need to revisit once all tests to see if every test is really `critical` ?

**Comment by sghai** [comment]
PR looks great.. just one small comment and I think we are good to merge then.

File: tests/foreman/api/test_template.py (hunk 8 above)
Comment (rochacbruno) [reply tests/foreman/api/test_template.py]: @sghai All template sync operations (import & export) will overwrite existing templates, so it is **critical** to have things like filters and subdirectories working.

Comment (sghai) [reply tests/foreman/api/test_template.py]: cool.. thanks

**Comment by sghai** [comment]
ACK
